### PR TITLE
New indicator: Active Fire Events (FIRMS)

### PR DESCRIPTION
**Pull request by Ohm-Np**
Resource: 
**nasa_firms**

Indicators:
 **active_fire_counts** and **active_fire_properties**

### CODE REVIEW

**Comment by Ohm-Np** [comment]
The routines workflow:

1. Users specify the instrument (satellite) as argument in get function:
    - **MODIS** (1 km resolution) - 2000 to 2021
    - **VIIRS S-NPP** (375 m resolution) - 2012 to 2021

 2. Get function:
    - downloads the gadm gpkg (world)
    - loads gadm polygons with wkt filter
    - extract iso3 code from gadm polygons
    - match iso3 code with country name provided by the txt file 'countriesISO.csv' 
    - get urls based on target years, country name and instrument defined
    - downloads CSVs and convert CSVs to gpkg file
    - returns nasa firms gpkg

3. Calc active fire counts:
   - intersects nasa_firms gpkg with portfolio object
   - returns tibble with number of fire events and corresponding year

4. Calc active fire properties:
    - intersects nasa_firms gpkg with portfolio object
    - returns tibble with all the variables along with lat/lon coordinates of fire events

**Comment by Jo-Schie** [comment]
ah okay. here it is better for discussion. @goergen95 . Should I test anything or look at the files or how do we organize this?

**Comment by Jo-Schie** [comment]
> The routines workflow:
> 
> 
> 
> 1. Users specify the instrument (satellite) as argument in get function:
> 
>     - **MODIS** (1 km resolution) - 2000 to 2021
> 
>     - **VIIRS S-NPP** (375 m resolution) - 2012 to 2021 (this one is default value)
> 
> 
> 
>  2. Get function:
> 
>     - downloads the gadm gpkg (world)
> 
>     - loads gadm polygons with wkt filter
> 
>     - extract iso3 code from gadm polygons
> 
>     - match iso3 code with country name provided by the txt file 'countriesISO.csv' 
> 
>     - get urls based on target years, country name and instrument defined
> 
>     - downloads CSVs and convert CSVs to gpkg file
> 
>     - returns nasa firms gpkg
> 
> 
> 
> 3. Calc active fire counts:
> 
>    - intersects nasa_firms gpkg with portfolio object
> 
>    - returns tibble with number of fire events and corresponding year
> 
> 
> 
> 4. Calc active fire properties:
> 
>     - intersects nasa_firms gpkg with portfolio object
> 
>     - returns tibble with all the variables along with lat/lon coordinates of fire events

So what is actually done with the GADM data? Is it used to get the spatial bounding box? I think this step is somehow missing in your description...

**Comment by Ohm-Np** [comment]
> what is actually done with the GADM data? 

The CSVs in the source-site are named and stored with country name like this `Dominican_Republic.csv`. And it might be cumbersome for users to first find the countries where the portfolio lies and then pass it as argument.

So, GADM data is used here to automatize this step: just only to extract country name/s.

**Comment by goergen95** [comment]
Hi, thanks for your work on the new indicator. I have some issues with the download and storing of the resource. First of all, the GADM data is about ~1 GB download and it is downloaded every single time a user requests the resource which I would like to avoid. Secondly, with the current approach, if the `nasa_firms` data directory exists, we will automatically append the  `active_fire.gpkg` file as a resource with no way of checking if the spatial extent actually matches the portfolio.

To avoid both I would thus suggest to download the global zip files from [here](https://firms.modaps.eosdis.nasa.gov/country/) and merge them together as GPKG per instrument and year (i.e. something linke `active_fire_viirs_2016.gpkg`). These files are usually below 400 MB in size to download and once the GPKGs are written to disk we can easily check if we need to download additional files with the user-supplied portfolio.

**Comment by Jo-Schie** [comment]
> > what is actually done with the GADM data? 
> 
> 
> 
> The CSVs in the source-site are named and stored with country name like this `Dominican_Republic.csv`. And it might be cumbersome for users to first find the countries where the portfolio lies and then pass it as argument.
> 
> 
> 
> So, GADM data is used here to automatize this step: just only to extract country name/s.
> 
> 
> 
> Since it all happens within the get functions, later we delete GADM data and CSVs as well within the functions and keep only the active fire polygons that are within the portfolio extent.

Thanks for clarification. So if I understand correctly, GADM data is intersected with country polygon boundaries to get the country names that then serve for fire data download...correct? 

If that is correct, then we should look for another source that is neither 1GB not 400MB IMHO. There is, for example, the world simple dataset in maptools which has only 6.4 Mb storage space and also comes with iso3 codes etc. we could just copy that from their Repo and add it to our package assets. What do you think @goergen95 & @Ohm-Np ?

**Comment by goergen95** [comment]
I am hesitant to include this in the package. The geoinformation is there, so there is actually no need to query some ISO strings once the data has been transformed to a spatial format. We also download other global resources (think TEOW) and even larger datasets like GFW so I don't see an issue with stitching a global firms GPKG in the get_resource function. Also, we have to make sure in some way that if a user changes his/her portfolio we make sure that the extents match and if not we download additional data. I don't see this happening with the other approaches...

**Comment by Jo-Schie** [comment]
> I am hesitant to include this in the package. The geoinformation is there, so there is actually no need to query some ISO strings once the data has been transformed to a spatial format. We also download other global resources (think TEOW) and even larger datasets like GFW so I don't see an issue with stitching a global firms GPKG in the get_resource function. Also, we have to make sure in some way that if a user changes his/her portfolio we make sure that the extents match and if not we download additional data. I don't see this happening with the other approaches...

I think the problem was that the data is to large if you download it for the whole globe correct? Maybe I totally do not get the discussion right, but I thought, that subsetting for countries was necessary in order to avoid downloading GBs and GBs of firms data...?

**Comment by Jo-Schie** [comment]
or let me rephrase: The data that you are suggesting for download @goergen95 . Is it exactly the same that @Ohm-Np is downloading in his routine? I just checked and I think the global datasets would be totally fine. nevertheless I am unsure if it is really the same data or whether the data from @Ohm-Np has some higher temporal resolution or something.

**Comment by Ohm-Np** [comment]
> the only difference that yours is global and the other per country

The routine currently works on individual CSV files - based on country name but the data Darius suggested are zipped as global - one time download - but within that zip file it consists hundreds ~200 CSVs named by their country. So, the source is same, the only difference is downloading single CSV file at a time vs all files at once.

**Comment by Ohm-Np** [comment]
So @goergen95,

as far as I understood:

1. download the global zipped files (per year, per instrument)
2. unzip and bind all CSVs to one
3. convert to geo-package
4. and save as `active_fire_viirs_2016.gpkg` or `active_fire_modis_2016.gpkg`

Did I get it correct?

**Comment by Jo-Schie** [comment]
> > the only difference that yours is global and the other per country
> 
> The routine currently works on individual CSV files - based on country name but the data Darius suggested are zipped as global - one time download - but within that zip file it consists hundreds ~200 CSVs named by their country. So, the source is same, the only difference is downloading single CSV file at a time vs all files at once and doing so, we avoid the necessity to involve GADM routine here.

in that case I favour @goergen95 approach as well. makes it also easier to maintain and is safer in terms of matching the data in the end. Regarding efficiency though I am a bit skeptical. The routine will need to intersect then potentially millions of points with a couple of AOIs. maybe you should test that first, if it works at all for one example data and maybe year 2015...

**Comment by Ohm-Np** [comment]
> download the global zipped files (per year, per instrument)
unzip and bind all CSVs to one
convert to geo-package
and save as active_fire_viirs_2016.gpkg or active_fire_modis_2016.gpkg

Doing so creates around 3.2 GB of file for a single year and instrument.

**Comment by goergen95** [comment]
ok, that is not really an issue though we could look for the most efficient way to create these geopackages (e.g. by appending instead of creating one large sf object before writing to disk). Could you push your code so far? I would be interested in testing the indicator calculation routine. I expect the `processing_mode = asset` to work more efficiently but that is something we should clarify.

**Comment by Ohm-Np** [comment]
The Ubuntu WSL on my PC is not responding now, so I could not test the package, but I have pushed the latest changes from my side and the new routines work fine at least for me.

**Comment by Jo-Schie** [comment]
Hi @goergen95 and @Ohm-Np . Is there anything missing for this pull request?  or can we merge? I'd like to test.

**Comment by Ohm-Np** [comment]
> Is there anything missing for this pull request?

some small updates on vignettes - _list of indicators/resources_

**Comment by goergen95** [comment]
Besides a small update to the NEWS.md I think we are ready for merging this PR!
